packages/block-editor/src/components/block-navigation/use-block-navigation-drop-zone.js

```diff
@@ -0,0 +1,309 @@
+ */
+
+/**
+ * An array representing data for blocks in the DOM used by drag and drop.
+ *
+ * @typedef {Object} WPBlockNavigationDropZoneBlocks
+ * @property {string}  clientId                        The client id for the block.
+ * @property {string}  rootClientId                    The root client id for the block.
+ * @property {number}  blockIndex                      The block's index.
+ * @property {Element} element                         The DOM element representing the block.
+ * @property {number}  innerBlockCount                 The number of inner blocks the block has.
+ * @property {boolean} isDraggedBlock                  Whether the block is currently being dragged.
+ * @property {boolean} canInsertDraggedBlocksAsSibling Whether the dragged block can be a sibling of this block.
+ * @property {boolean} canInsertDraggedBlocksAsChild   Whether the dragged block can be a child of this block.
+ */
+
+/**
+ * An object containing details of a drop target.
+ *
+ * @typedef {Object} WPBlockNavigationDropZoneTarget
+ * @property {string}                   blockIndex   The insertion index.
+ * @property {string}                   rootClientId The root client id for the block.
+ * @property {string|undefined}         clientId     The client id for the block.
+ * @property {'top'|'bottom'|'inside'}  dropPosition The position relative to the block that the user is dropping to.
+ *                                                   'inside' refers to nesting as an inner block.
+ */
+
+/**
+ * A react hook that returns data about blocks used for computing where a user
+ * can drop to when dragging and dropping blocks.
+ *
+ * @param {Object}          ref           A React ref of a containing element for block navigation.
+ * @param {WPPoint}         position      The current drag position.
+ * @param {WPDragEventType} dragEventType The drag event type.
+ *
+ * @return {WPBlockNavigationDropZoneBlocks} An array representing data for each block in the block navigation DOM.
+ */
+function useDropTargetBlocksData( ref, position, dragEventType ) {
+	const {
+		getBlockRootClientId,
+		getBlockIndex,
+		getBlockCount,
+		getDraggedBlockClientIds,
+		canInsertBlocks,
+	} = useSelect( ( select ) => {
+		const {
+			canInsertBlocks: _canInsertBlocks,
+			getBlockRootClientId: _getBlockRootClientId,
+			getBlockIndex: _getBlockIndex,
+			getBlockCount: _getBlockCount,
+			getDraggedBlockClientIds: _getDraggedBlockClientIds,
+		} = select( 'core/block-editor' );
+		return {
+			canInsertBlocks: _canInsertBlocks,
+			getBlockRootClientId: _getBlockRootClientId,
+			getBlockIndex: _getBlockIndex,
+			getBlockCount: _getBlockCount,
+			getDraggedBlockClientIds: _getDraggedBlockClientIds,
+		};
+	}, [] );
+
+	// Compute data about blocks only when the user
+	// starts dragging, as determined by `hasPosition`.
+	const hasPosition = !! position;
+	return useMemo( () => {
+		if ( ! ref.current || ! hasPosition ) {
```

Comment:
If the `ref` updated, this hook will never be triggered again, and the memo will never be recalculated. Not a big deal if the `ref` would never update, but would potentially introduce bugs when someone decide to use this hook at difference places.

Maybe just make this hook accept the `HTMLElement` type as the first argument? So that the existing component can just pass `ref.current` to this hook. And more advance usage can use ref callback and state to trigger recalculation to this hook.

Reply:
Hmm, good spot. Originally I had this as an effect, so it was controllable using just the `hasPosition` dependency, but I've incorrectly assumed this could be converted to `useMemo` without changing that. Is there an issue with changing the `useMemo` dependencies? e.g.:
```jsx
	return useMemo( () => {
            // snip ....
	}, [ ref.current, hasPosition, dragEventType ] );
```

I would kind of prefer `ref` parameters, since `useDropZone` already accepts a `ref` (even though the property is called `element` it is still a ref).

Reply:
Actually, I think I should change it back to an effect to be safest. It should really recompute on every drag since the child elements of `ref.current` might change, and referential equality won't cover that.

Reply:
Done in 575aa75c61e835d9f81c114aa3974914fdf5a7f4

Reply:
Awesome!

Just a side note that we tend to not putting `ref.current` in the hook dependency array, because it just doesn't help at all.

Updating `ref.current` won't trigger a re-render, so putting it inside dependency array doesn't make sense.

I believe there is also a rule in the official eslint hook plugin that catches exactly this kind of issue. I thought that we had it enabled 🤔 ?